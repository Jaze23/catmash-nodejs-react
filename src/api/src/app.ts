import "reflect-metadata";
import { env } from "./common/config";
import express from "express";
import cors from "cors";
import helmet from "helmet";
import bodyParser from "body-parser";
import { connect } from "mongoose";
import mongooseSchemaLoader from "./common/config/mongoose-schema-loader";
import routerV1 from "./routers/router-v1";
import { errorHandlerMiddleware, cls } from "./common/middlewares";

const app = express();

if (process.env.NODE_ENV !== "test") {
  connect(
    env.database.url as string,
    { useNewUrlParser: true, useCreateIndex: true },
  );
}

// Load schemas into DI container
mongooseSchemaLoader();

// Add CORS
app.use(cors({ origin: "*" })); // should be set to only allowed values.

// Add helmet
app.use(helmet());

// Configure the express app
app.set("port", process.env.PORT || 8080);
app.use(bodyParser.json({ limit: "50MB" }));
app.use(bodyParser.urlencoded({ extended: true }));

// Add the CLS
app.use(cls);

// Register the router
app.use(`/v1`, routerV1);

// Add the error handler
app.use(errorHandlerMiddleware);

export default app;
