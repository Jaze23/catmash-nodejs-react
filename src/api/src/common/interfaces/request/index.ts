export * from "./typed-request.interface";
