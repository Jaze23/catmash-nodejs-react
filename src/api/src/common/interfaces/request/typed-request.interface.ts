export interface TypedRequest<T> {
  body: T;
}
