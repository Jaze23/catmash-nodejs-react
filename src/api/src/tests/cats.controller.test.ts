import request from "supertest";
import app from "../app";
import { SetupTest, createCats, getACat } from "./setup/setup";

const setupTest = new SetupTest();

beforeEach(async done => {
  await setupTest.initAndConnectToMongodbTest();
  await createCats();
  done();
});

describe("GET /cats", () => {
  it("should return cats", async () => {
    const response = await request(app).get(`/v1/cats`);

    expect(response.status).toBe(200);
    expect(response.body[0]).toMatchSnapshot({
      _id: expect.any(String),
      id: expect.any(String),
      url: expect.any(String),
    });
  });
});

describe("GET /cats/versus", () => {
  it("should return 2 cats", async () => {
    const response = await request(app).get(`/v1/cats/versus`);

    expect(response.status).toBe(200);
    expect(response.body.length).toBe(2);
    expect(response.body[0]).toMatchSnapshot({
      _id: expect.any(String),
      id: expect.any(String),
      url: expect.any(String),
    });
  });
});

describe("PUT /cats/:id/vote", () => {
  it("should add a vote to the cat", async () => {
    const cat = await getACat();
    const response = await request(app).put(`/v1/cats/${cat._id}/vote`);

    expect(response.status).toBe(200);
    expect(response.body).toMatchSnapshot({
      _id: expect.any(String),
      id: expect.any(String),
      url: expect.any(String),
      votes: cat.votes + 1,
    });
  });

  it("should a 500 if the requested id is not a valid ObjectId", async () => {
    const cat = await getACat();
    const response = await request(app).put(`/v1/cats/nonexistingid/vote`);

    expect(response.status).toBe(500);
    expect(response.body).toMatchSnapshot({});
  });

  it("should a 404 if the requested id does not exist", async () => {
    const cat = await getACat();
    const response = await request(app).put(
      `/v1/cats/507f191e810c19729de860ea/vote`,
    );

    expect(response.status).toBe(404);
    expect(response.body).toMatchSnapshot({});
  });
});

afterEach(async done => {
  await setupTest.disconnectAndStopMongodb();
  done();
});
