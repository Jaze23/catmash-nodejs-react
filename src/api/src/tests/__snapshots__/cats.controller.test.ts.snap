// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /cats should return cats 1`] = `
Object {
  "_id": Any<String>,
  "id": Any<String>,
  "url": Any<String>,
  "votes": 0,
}
`;

exports[`GET /cats/versus should return 2 cats 1`] = `
Object {
  "_id": Any<String>,
  "id": Any<String>,
  "url": Any<String>,
  "votes": 0,
}
`;

exports[`PUT /cats/:id/vote should a 404 if the requested id does not exist 1`] = `
Object {
  "message": "Cat not found",
  "status": 404,
}
`;

exports[`PUT /cats/:id/vote should a 500 if the requested id is not a valid ObjectId 1`] = `
Object {
  "message": "Something went wrong",
  "status": 500,
}
`;

exports[`PUT /cats/:id/vote should add a vote to the cat 1`] = `
Object {
  "_id": Any<String>,
  "id": Any<String>,
  "url": Any<String>,
  "votes": 1,
}
`;
